pkg/logentry/stages/pipeline.go

```diff
@@ -7,37 +7,46 @@ import (
 	"github.com/go-kit/kit/log/level"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/prometheus/common/model"
 
-	"github.com/grafana/loki/pkg/logentry/stages"
 	"github.com/grafana/loki/pkg/promtail/api"
 )
 
-var (
-	pipelineDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
-		Namespace: "logentry",
-		Name:      "pipeline_duration_seconds",
-		Help:      "Label and metric extraction pipeline processing time, in seconds",
-		Buckets:   []float64{.000005, .000010, .000025, .000050, .000100, .000250, .000500, .001000, .002500, .005000, .010000, .025000},
-	}, []string{"job_name"})
-)
-
 // PipelineStages contains configuration for each stage within a pipeline
-type PipelineStages []interface{}
+type PipelineStages = []interface{}
+
+// PipelineStage contains configuration for a single pipeline stage
+type PipelineStage = map[interface{}]interface{}
 
 // Pipeline pass down a log entry to each stage for mutation and/or label extraction.
 type Pipeline struct {
-	logger  log.Logger
-	stages  []stages.Stage
-	jobName string
+	logger     log.Logger
+	stages     []Stage
+	jobName    *string
+	plDuration *prometheus.HistogramVec
 }
 
 // NewPipeline creates a new log entry pipeline from a configuration
-func NewPipeline(logger log.Logger, stgs PipelineStages, jobName string) (*Pipeline, error) {
-	st := []stages.Stage{}
+func NewPipeline(logger log.Logger, stgs PipelineStages, jobName *string, registerer prometheus.Registerer) (*Pipeline, error) {
+	hist := prometheus.NewHistogramVec(prometheus.HistogramOpts{
```

Comment:
I think this is safe to keep this the default registerer at init time. I introduced the registerer if you wanted to have your custom metrics exposed differently which this one is not custom, it's a promtail one.

At the end of the day the registration will work for only the first one, then the rest will used the same registered metrics.